### PR TITLE
feat(artifacts): Introduce ArtifactPublisher plugin interface (2 of 3)

### PR DESCRIPTION
**Pull request by luispollo**
Follow up to #1304.

This PR is functionally a no-op, just like the previous one, and focuses on moving files around into the proper places in preparation to remove the dependency on `kork-artifacts` and implement the `ArtifactPublisher` plugin interface for Docker and Debian in keel in the next PR.

~CI builds for this PR will break until #1304 is merged.~

### CODE REVIEW

File: keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt [hunk 1]
```diff
@@ -55,13 +50,6 @@ object KeelApiModule : SimpleModule("Keel API") {
       setMixInAnnotations<Monikered, MonikeredMixin>()
       setMixInAnnotations<SubnetAwareRegionSpec, SubnetAwareRegionSpecMixin>()
       setMixInAnnotations<ConstraintState, ConstraintStateMixin>()
-
-      registerSubtypes(
```
Comment (luispollo) [inline keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt:59]: This is the only functional change in this PR. I did it to avoid a circular dependency between `keel-core` and `keel-artifact` via this module.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/VersioningStrategy.kt [hunk 2]
```diff
@@ -3,7 +3,7 @@ package com.netflix.spinnaker.keel.api.artifacts
 /**
  * Strategy for how to sort versions of artifacts.
  */
-sealed class VersioningStrategy
+abstract class VersioningStrategy
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/VersioningStrategy.kt:6]: So they can be implemented in plugins.

File: keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifact/events/KorkArtifactEvent.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.netflix.spinnaker.keel.artifact.events
+
+import com.netflix.spinnaker.kork.artifacts.model.Artifact
+
+data class KorkArtifactEvent(
```
Comment (luispollo) [inline keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifact/events/KorkArtifactEvent.kt:5]: This is a temporary replacement for `ArtifactEvent` which still uses a kork `Artifact` instead of the plugin-friendly `SpinnakerArtifact`. It will be removed in part 3.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/VersioningStrategyComparators.kt [hunk 4]
```diff
@@ -31,9 +31,11 @@ import org.slf4j.LoggerFactory
 import org.springframework.util.comparator.NullSafeComparator
 
 val VersioningStrategy.comparator: Comparator<String>
+  // TODO: this needs to support versioning strategies provided by artifact plugins
```
Comment (lorin) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/VersioningStrategyComparators.kt:34]: Is this TODO going to be implemented in part 3, or is that functionality not coming until later?

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/ArtifactUtils.kt [hunk 5]
```diff
@@ -26,16 +26,18 @@ import com.netflix.spinnaker.keel.api.artifacts.ArtifactStatus
...
 import com.netflix.spinnaker.keel.api.artifacts.VirtualMachineOptions
+import com.netflix.spinnaker.keel.artifacts.DebianArtifact
+import com.netflix.spinnaker.keel.artifacts.DockerArtifact
 import com.netflix.spinnaker.keel.serialization.configuredObjectMapper
 
 private val objectMapper: ObjectMapper = configuredObjectMapper()
 
 /**
  * A helper function to construct the proper artifact type from the serialized json.
+ * FIXME: this needs to go away in favor of [ArtifactPublisher] functions to keep
```
Comment (lorin) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/ArtifactUtils.kt:39]: Is this coming in PR 3?

File: keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt (hunk 1 above)
Comment (emjburns) [reply keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt:59]: Functional change as in, when this pr is merged then this functionality is removed?

File: keel-web/keel-web.gradle.kts [hunk 6]
```diff
@@ -21,6 +21,7 @@ dependencies {
   implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core")
   implementation("com.netflix.spinnaker.kork:kork-core")
   implementation("com.netflix.spinnaker.kork:kork-web")
+  implementation("com.netflix.spinnaker.kork:kork-artifacts")
```
Comment (emjburns) [inline keel-web/keel-web.gradle.kts:24]: I thought you were trying to remove the `kork-artifacts` dependency?

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt [hunk 7]
```diff
@@ -1,53 +1,36 @@
...
+
+  override fun toString() = name
 }
 
 enum class ArtifactStatus {
   FINAL, CANDIDATE, SNAPSHOT, RELEASE, UNKNOWN
 }
 
-sealed class DeliveryArtifact {
+/**
```
Comment (emjburns) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt:13]: Why did you move these things to the `DeliveryArtifact` class? I feel like if you're going to move them they might as well be moved into their own classes.

File: keel-artifact/keel-artifact.gradle.kts [hunk 8]
```diff
@@ -9,7 +9,7 @@ dependencies {
   implementation(project(":keel-clouddriver"))
   implementation("org.springframework:spring-context")
   implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core")
-
+  implementation("com.netflix.spinnaker.kork:kork-artifacts")
```
Comment (emjburns) [inline keel-artifact/keel-artifact.gradle.kts:12]: I thought you were trying to remove the dependency on `kork-artifacts`?

File: keel-core/keel-core.gradle.kts [hunk 9]
```diff
@@ -35,6 +34,7 @@ dependencies {
   api("net.swiftzer.semver:semver:1.1.0")
   api("org.jetbrains.kotlinx:kotlinx-coroutines-core")
   api("org.jetbrains.kotlinx:kotlinx-coroutines-slf4j")
```
Comment (emjburns) [inline keel-core/keel-core.gradle.kts:36]: Why did you add this commons lang dependency? I don't see it used anywhere, unless I am missing something

File: keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt (hunk 1 above)
Comment (luispollo) [reply keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt:59]: Yes, I just realized I can't merge this as is because it would break serialization of these sub-types until they're registered some other way. I'll think of something else.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/VersioningStrategyComparators.kt (hunk 4 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/VersioningStrategyComparators.kt:34]: Part 3.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/ArtifactUtils.kt (hunk 5 above)
Comment (luispollo) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/ArtifactUtils.kt:39]: I'm trying very hard to fix it for part 3, yes. We'll see... 😅

File: keel-web/keel-web.gradle.kts (hunk 6 above)
Comment (luispollo) [reply keel-web/keel-web.gradle.kts:24]: Yes, this is only temporary while kork's `Artifact` coexists with the new `SpinnakerArtifact` in this PR.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt (hunk 7 above)
Comment (luispollo) [reply keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt:13]: Fair point. Will do.

File: keel-artifact/keel-artifact.gradle.kts (hunk 8 above)
Comment (luispollo) [reply keel-artifact/keel-artifact.gradle.kts:12]: Yes, this is only temporary while kork's Artifact coexists with the new SpinnakerArtifact in this PR.

File: keel-core/keel-core.gradle.kts (hunk 9 above)
Comment (luispollo) [reply keel-core/keel-core.gradle.kts:36]: It was coming in with `kork-artifacts`. If you search for `import org.apache.commons.lang3` you'll find lots of usages.

File: keel-web/keel-web.gradle.kts (hunk 6 above)
Comment (emjburns) [reply keel-web/keel-web.gradle.kts:24]: Ah, okay, got it.

File: keel-core/keel-core.gradle.kts (hunk 9 above)
Comment (emjburns) [reply keel-core/keel-core.gradle.kts:36]: ah, ok, got it! so I was missing something :D

File: keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt (hunk 1 above)
Comment (lorin) [reply keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt:59]: Based on this comment, I'm going to change approval status of my review.

Comment (luispollo) [reply keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt]: I've reverted that part of the PR. Please re-review.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt [hunk 10]
```diff
@@ -1,53 +1,20 @@
...
+
+  override fun toString() = name
 }
 
 enum class ArtifactStatus {
   FINAL, CANDIDATE, SNAPSHOT, RELEASE, UNKNOWN
 }
 
-sealed class DeliveryArtifact {
+abstract class DeliveryArtifact {
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt:13]: So it can be extended by plugins.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt [hunk 11]
```diff
@@ -1,53 +1,20 @@
...
   abstract val name: String
   abstract val type: ArtifactType
   abstract val versioningStrategy: VersioningStrategy
   abstract val reference: String // friendly reference to use within a delivery config
   abstract val deliveryConfigName: String? // the delivery config this artifact is a part of
-  override fun toString() = "$type:$name" + if (reference != null) " (ref: $reference)" else ""
-}
-
-data class DebianArtifact(
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt:25]: Moved to its own class outside the `api` package so it can later be moved out of `keel-api` into another module.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt [hunk 12]
```diff
@@ -1,53 +1,20 @@
...
-  override val deliveryConfigName: String? = null,
-  override val reference: String = name,
-  val vmOptions: VirtualMachineOptions,
-  val statuses: Set<ArtifactStatus> = emptySet(),
-  override val versioningStrategy: VersioningStrategy = DebianSemVerVersioningStrategy
-) : DeliveryArtifact() {
-  override val type = ArtifactType.deb
-}
-
-data class DockerArtifact(
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/artifacts/DeliveryArtifact.kt:36]: Moved to its own class outside the api package so it can later be moved out of keel-api into another module.

File: keel-ec2-plugin/keel-ec2-plugin.gradle.kts [hunk 13]
```diff
@@ -5,11 +5,12 @@ plugins {
 
 dependencies {
   api(project(":keel-api"))
-  implementation(project(":keel-api-jackson"))
```
Comment (emjburns) [inline keel-ec2-plugin/keel-ec2-plugin.gradle.kts:8]: should this just be removed?

**Comment by luispollo** [comment]
@lorin @emjburns Updated based on conversations above. Please take another 👀  when you have a chance! Thanks!
